sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2146,7 +2146,7 @@ class Dataset[T] private[sql](
    * Returns a new Dataset by adding columns or replacing the existing columns that has
    * the same names.
    */
-  private[spark] def withColumns(colNames: Seq[String], cols: Seq[Column]): DataFrame = {
+  private[spark] def withColumns(colNames: Seq[String], cols: Seq[Column]): DataFrame = withPlan {
```

Comment:
As stated on the JIRA ticket, the problem is deep query plan. I think we can have many ways to create such deep query plan, not only for `withColumns`. For example, you can call `select` many times to do that too. This change makes `withColumns` a special case.

Reply:
yes, but I think this is a special case. I have seen many cases when `withColumn` is used in for loops: with this change such a pattern would be better supported.